### PR TITLE
Migrated to cryptonite implementation of Argon2

### PR DESCRIPTION
**Pull request by conor10**
The recent cryptonite v0.22 release added Argon2 support.

https://hackage.haskell.org/package/cryptonite-0.22/docs/Crypto-KDF-Argon2.html

### CODE REVIEW

**Comment by patrickmn** [comment]
This is great. Thank you!

I do have to ask you to submit a CLA prior to merging this. Please provide an email address where you'd like to receive it or send a quick email with CLA in the subject to quorum_info@jpmorgan.com

File: Constellation/Util/Lockable.hs [hunk 1]
```diff
@@ -5,13 +5,14 @@
...
 
 import ClassyPrelude hiding (hash)
-import Crypto.Argon2
-    ( HashOptions( HashOptions, hashIterations, hashMemory, hashParallelism
-                 , hashVariant
-                 )
-    , Argon2Variant(Argon2d, Argon2i)
+import Crypto.Error (CryptoFailable(..))
+import Crypto.KDF.Argon2
+    ( Options(..)
```
Comment (patrickmn) [inline Constellation/Util/Lockable.hs]: These should probably be one line

File: Constellation/Util/Lockable.hs [hunk 2]
```diff
@@ -80,33 +82,48 @@ instance FromJSON ArgonSalt where
         Right asalt -> return $ ArgonSalt asalt
     parseJSON _             = fail "ArgonSalt must be an Aeson String"
 
-newtype ArgonOptions = ArgonOptions { unArgonOptions :: HashOptions }
+newtype ArgonOptions = ArgonOptions { unArgonOptions :: Options }
+instance Show ArgonOptions where
```
Comment (patrickmn) [inline Constellation/Util/Lockable.hs]: \n before

File: Constellation/Util/Lockable.hs [hunk 3]
```diff
@@ -80,33 +82,48 @@ instance FromJSON ArgonSalt where
...
+    show (ArgonOptions options) = show options
+instance Eq ArgonOptions where
```
Comment (patrickmn) [inline Constellation/Util/Lockable.hs]: \n before

File: stack.yaml [hunk 4]
```diff
@@ -73,4 +72,4 @@ extra-package-dbs: []
 
 #build:
 #  library-profiling: true
-#  executable-profiling: true
+#  executable-profiling: true
```
Comment (patrickmn) [inline stack.yaml]: missing \n before EOF

File: stack.yaml [hunk 5]
```diff
@@ -37,15 +37,14 @@ resolver: lts-8.3
...
-- location: https://github.com/jpmorganchase/constellation/files/590600/argon2-1.3.0.tar.gz
-  extra-dep: true
 
 # Dependency packages to be pulled from upstream that are not in the resolver
 # (e.g., acme-missiles-0.3)
 extra-deps:
   - BerkeleyDB-0.8.7
   - logging-3.0.4
   - saltine-0.0.0.4
+  - cryptonite-0.22
```
Comment (patrickmn) [inline stack.yaml]: should be between BerkeleyDB and logging

File: test/Constellation/Util/Lockable/Test.hs [hunk 6]
```diff
@@ -2,9 +2,61 @@
 {-# LANGUAGE OverloadedStrings #-}
 module Constellation.Util.Lockable.Test where
 
+import Data.Tuple (swap)
```
Comment (patrickmn) [inline test/Constellation/Util/Lockable/Test.hs]: These imports should be after ClassyPrelude

File: test/Constellation/Util/Lockable/Test.hs [hunk 7]
```diff
@@ -2,9 +2,61 @@
...
+import qualified Data.Text.Lazy as TL
+import qualified Data.Text.Lazy.Encoding as TLE
+
 import ClassyPrelude
 import Test.Tasty (TestTree, testGroup)
 import Test.Tasty.HUnit ((@?=), testCase)
+import Data.Aeson (encode, decode)
```
Comment (patrickmn) [inline test/Constellation/Util/Lockable/Test.hs]: Should be after ClassyPrelude

File: test/Constellation/Util/Lockable/Test.hs [hunk 8]
```diff
@@ -2,9 +2,61 @@
...
+
+import Constellation.TestUtil (kvTest)
+import Constellation.Util.Lockable
+    ( ArgonOptions (ArgonOptions)
```
Comment (patrickmn) [inline test/Constellation/Util/Lockable/Test.hs]: could be one line

File: test/Constellation/Util/Lockable/Test.hs [hunk 9]
```diff
@@ -2,9 +2,61 @@
...
+    , defaultArgonOptions 
+    )
+import Crypto.KDF.Argon2
+    ( Options(..)
```
Comment (patrickmn) [inline test/Constellation/Util/Lockable/Test.hs]: could be one line

**Comment by patrickmn** [comment]
In the test data, is the NoVersion stub data a key that was generated by the previous version? (So we know that it can still unpack existing keys.)

**Comment by conor10** [comment]
Formatting changes have been made.

The NoVersion stub is the same as the defaultArgonOptions with the version field removed.

**Comment by patrickmn** [comment]
Thanks for the email. I uploaded the CLA here for convenience: https://github.com/jpmorganchase/cla/blob/master/cla-2017-09-01.pdf

Please sign and scan, then email it to the same address.

**Comment by patrickmn** [comment]
The only thing I'd like to add is a test unlocking this key that was generated pre-change (but with much lower overhead than usual) with password "foo":

    {"data":{"aopts":{"variant":"i","memory":8192,"iterations":2,"parallelism":2},"snonce":"LJ7QUiZ0kD/mPf/aHvCEQKDhlUucZM0s","asalt":"I5bvdFhNpr8hlknzOHIO7Q3S5Ubs9ucUdVneghrebZQ=","sbox":"iHJAfcpWWE5PsvwZRwr5gCtwMKu7xXZAAPCVRGfx06l/I5QgXs0V/czOqYsMxoEG"},"type":"argon2sbox"}

(I'm happy to do this.)

**Comment by conor10** [comment]
Sent. 

Sure - makes sense to have serialisation test coverage of more of the Lockable types.

**Comment by patrickmn** [comment]
For good measure: https://github.com/jpmorganchase/constellation/commit/f4f354a3309482bc836f819e6c3b4d688442be6e

Thanks again

**Comment by cartazio** [comment]
good job :)

**Pull request merged**